src/python/pants/testutil/rule_runner.py

```diff
@@ -793,23 +877,43 @@ def run_rule_with_mocks(
     if not isinstance(res, (Coroutine, Generator)):
         return res
 
-    def get(res: Get | Effect | Call):
+    def get(res: Get | Effect | Call | Coroutine):
+        if isinstance(res, Coroutine):
+            # A call-by-name element in a concurrently() is a Coroutine whose frame is
+            # the trampoline wrapper that creates and immediately awaits the Call.
+            assert res.cr_frame is not None
+            rule_id = res.cr_frame.f_locals["rule_id"]
+            args = res.cr_frame.f_locals["args"]
+            mock_call = mock_calls.get(rule_id)
+            if mock_call:
+                return mock_call(*args)
+            raise AssertionError(f"No mock_call provided for {rule_id}.")
+        elif isinstance(res, Call):
+            mock_call = mock_calls.get(res.rule_id)
+            if mock_call:
+                return mock_call(*res.inputs)
+            # For now we fall through, to allow an old-style MockGet to mock a call-by-name, for
+            # legacy reasons. But we will deprecate and then remove this in the future, at which
+            # point we should error here, e.g.,:
+            # raise AssertionError(f"No mock_call provided for {res.rule_id}.")
+            # Note that this fallthrough only works for single call-by-names. When wrapped in a
+            # concurrently() call, mock_calls *must* be used.
+
         provider = next(
             (
                 mock_get.mock
                 for mock_get in mock_gets
                 if mock_get.output_type == res.output_type
                 and all(
-                    type(val) in mock_get.input_types
+                    input_type in mock_get.input_types
                     or (
                         union_membership
+                        and input_type in union_membership
```

Comment:
The previous version of this union check was completely wrong: it got the union type/union member check backwards. It's not clear to me how the tests relying on this ever passed.